# COGS 108 - Audio Signatures of Cultural Identity in Musical Subgenre Evolution

# Permissions

Place an `X` in the appropriate bracket below to specify if you would like your group's project to be made available to the public. (Note that student names will be included (but PIDs will be scraped from any groups who include their PIDs).

* [] YES - make available
* [`X`] NO - keep private

# Names

- Pia Dihayco
- Asher Hardy
- Kelsie Lloyd
- Jiwoo Son
- Jason Tran

# Final Project Video

https://youtu.be/gwi8t9ipUX4

# Abstract

The aim of our project is to explore cultural music trends by analyzing various audio characteristics such as danceability, energy, loudness, acousticness, and tempo across different genres and countries. Based on data from Spotify, we created visualizations such as a heatmap to examine the correlation between these audio features and identified how they differed by genre and region. Positive correlations, like those between energy and loudness, highlighted patterns of high energy and loudness in electronic and energetic genres like EDM. On the contrary, negative correlations like between energy and acousticness, revealed how higher energy tracks often use fewer acoustic instruments. Additionally, we focused on how genres like dance pop, EDM, and tropical house, which utilize electronic instruments, are increasingly common compared to traditional genres like rock and hip-hop.

Our findings revealed a tendency towards danceable and energetic tracks, particularly in electronic genres. Regional analysis of music distribution showed that countries like the United States, Brazil, and major European nations had higher genre diversity, possibly due to the broader availability of music on platforms like Spotify. Latin American countries like Mexico and Brazil stood out with high danceability and loudness, reflecting cultural tendencies toward lively, energetic music. In contrast, countries with a more classical or traditional musical background, like France, exhibited different audio characteristics. The geographical distribution, represented in a choropleth map, emphasized how regional preferences influence the popularity of certain genres, though data limitations and regional market access influenced the findings.

We also used the Kruskal-Wallis H-test to identify significant differences in audio characteristics across various countries. The results indicated notable differences in energy, loudness, and acousticness, with countries like Japan showing higher energy and loudness levels. However, there was no significant difference in tempo, suggesting that tempo preferences are relatively consistent across the countries examined. The study underscored the increasing global appeal of electronic music while highlighting how cultural and regional influences shape musical preferences. While this analysis offers valuable insights into cultural musical tendencies, it also highlights the importance of considering data biases and market limitations when interpreting global music trends.







# Research Question

What systematic changes in audio characteristics (tempo, energy, instrumentalness, etc.) occur when a genre of music is adopted by separate ethnic groups, and does each culture show consistent patterns in how they modify these features?

## Background and Prior Work

Music is a form of arts and entertainment composed of sounds, both vocal and instrumental. Human beings have enjoyed music for thousands of years and it is an essential aspect of many cultures around the world. Music has a plethora of purposes, including self-expression, passing down tradition, and creating intergenerational connections. Different cultures utilize unique music styles, varying in characteristics such as instrumentation, rhythm, melody and vocals. Musical genre categorization has created a system for both artists and listeners to understand the diversity of music and shared traits within a specific genre. Intercultural interactions throughout history enabled the evolution and spread of music genres across different cultures<a name="cite_ref-1"></a>[<sup>[1]</sup>](#cite_note-1). Music genres not only retain their cultural origins but also reflect globalization, for example, K-Pop blends both Korean and Western influence<a name="cite_ref-2"></a>[<sup>[2]</sup>](#cite_note-2). However, more research to understand systematic differences in a certain music genre across cultures that have adapted it as their own.

Ethnic groups are defined as a group of people who identify with common things such as language, cultural practices, and history<a name="cite_ref-3"></a>[<sup>[3]</sup>](#cite_note-3). Unlike race, ethnicity is not solely defined by physical traits, though ethnic groups may find commonalities in their physical features as well. Furthermore, ethnicity is not static. Groups can change over time through means such as migration, intermarriage or globalization<a name="cite_ref-4"></a>[<sup>[4]</sup>](#cite_note-4). Thus, there's no single definite definition for what makes an ethnic group.

<a name="cite_note-1"></a>1. [^](#cite_ref-1) <a href="https://inspireculture.org/cultural-awareness/arts-culture/cross-cultural-influence-in-music-a-journey-through-genres/">https://inspireculture.org/cultural-awareness/arts-culture/cross-cultural-influence-in-music-a-journey-through-genres/</a><br>
<a name="cite_note-2"></a>2. [^](#cite_ref-2) <a href="https://aclanthology.org/2020.emnlp-main.386.pdf">https://aclanthology.org/2020.emnlp-main.386.pdf</a><br>
<a name="cite_note-3"></a>3. [^](#cite_ref-3) <a href="https://diversityresources.com/difference-between-race-ethnicity/">https://diversityresources.com/difference-between-race-ethnicity/</a><br>
<a name="cite_note-4"></a>4. [^](#cite_ref-4) <a href="https://sociology.institute/sociological-theories-concepts/defining-ethnic-groups-concepts-perspectives/#google_vignette">https://sociology.institute/sociological-theories-concepts/defining-ethnic-groups-concepts-perspectives/</a>

# Hypothesis


We hypothesize that when music genres are adopted by different ethnic groups, they will display statistically significant differences in measurable audio characteristics (i.e. tempo, energy, instrumentalness), with each ethnic group showing consistent patterns on how they modify these features. More tangibly, we predict that the subgenres will exhibit quantifiable variations in their audio measurements that correlate with the ethnic background of the artists.

This hypothesis is based on several factors. Firstly, cultural differences in musical traditions (i.e. rhythmic patterns, scale systems, instrumental preferences) are very likely to influence how artists adapt and interpret genres. Secondly, geographic/linguistic variations between ethnic groups often lead to particularly unique musical expressions, even within the same “framework” of genre (i.e. how tempo and rhythm might be modified to better accommodate different language patterns, or how traditional instruments might be incorporated into modern genres). Our hypothesis is also informed by empirical observations of how genres like hip-hop, pop, and electronic music have evolved differently across various cultural contexts, suggesting systematic patterns in how musical elements are modified to align with local musical traditions.

# Data

## Data Overview

- Dataset #1
  - Dataset Name: "EveryNoise database with spotify features" (`last_fm.db`)
  - Link to the dataset: https://www.kaggle.com/datasets/cesaregarza/everynoise-database-with-spotify-features
  - Number of observations: 764,429
  - Number of variables: 2

The EveryNoise database contains comprehensive audio feature data for over 750,000 songs collected through the Spotify API, including key metrics like danceability, instrumentalness, acousticness, energy, tempo, and valence. These quantitative features serve as proxies for musical characteristics that can help differentiate between cultural interpretations of genres. The dataset also includes rich metadata about artists and genres, which will allow us to map songs to their cultural/ethnic origins by analyzing both the artist's country of origin and the cultural indicators present in genre names (e.g. "k-pop", "swedish metal").

To prepare this data for analysis, we'll most likely need to clean the genre labels to extract cultural/regional indicators (such as "korean", "swedish", "dutch", etc.), standardize artist location data, and potentially aggregate the audio features to identify patterns at the genre-culture level rather than individual song level.

The database itself is also egregiously large (approx. 600MB) which is a problem for execution times. We will most likely need to generate a subset of the data (i.e. a 10% random sample) to perform our analysis, and also convert it into a `.csv` file (instead of an SQLite database) for easier handling.

## Dataset #1: "EveryNoise database with Spotify features" (`last_fm.db`)

The blatant issue here is that the dataset is too large to even place in this GitHub repository and would need to be offshored in Git LFS. For demonstration purposes here, we will use an example SQLite database that is a 10% random sample of the full dataset (76,442 observations), and show how we would load/clean/wrangle this into a 10% random sample `.csv` file (7,644 observations). In reality, we will be simply taking a 10% random sample of the full dataset (meaning our final `.csv` would have 76,442 observations). The original dataset will not be included in this repository, but the sample dataset will be available at `data/last_fm_sample.db`, and the `.csv` file that we sampled from the real dataset (and thus will use for analysis) will be available at `data/last_fm.csv`.

We first import our necessary dependencies:

In [ ]:
import sqlite3
import pandas as pd
import numpy as np
import json
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
from scipy.stats import kruskal

We can establish a "connection" to our SQLite database using the `sqlite3` library with `.connect()`, and load the `FEATURES_2` and `TRAINING_DB_2` tables into pandas DataFrames with `.read_sql_query()`:

In [ ]:
connection = sqlite3.connect("data/last_fm_sample.db")

features_df = pd.read_sql_query("SELECT * FROM FEATURES_2", connection)
training_df = pd.read_sql_query("SELECT * FROM TRAINING_DB_2", connection)

display(features_df.head())
display(training_df.head())

,track_id,album_name,album_id,album_release_date,album_total_tracks,artist_name,artist_id,track_duration,track_name,explicit,...,key,loudness,mode,speechiness,instrumentalness,acousticness,liveness,valence,tempo,time_signature
0,2sDurGdooRkMxYRGfRvPdn,Hifalutin,4ko6ecNnYMliJ7Aon2ckdS,2019-04-12,14,Lowly,721bfjbiX9Koqt2rO8PWOY,267477,baglaens,0,...,10,-7.735,1,0.0399,0.317000,0.683000,0.1270,0.232,170.057,4
1,00iwom2ONo0LvxvRkeXQdD,Accordéons de France,3NgdAeFdTGYEsFqzqiI10O,2016-08-26,20,Karine Fontaine,5ge9YDw6vooNaGcXGKTaPn,158306,Rue Championnet,0,...,9,-2.614,1,0.0578,0.016900,0.417000,0.0822,0.822,187.702,3
2,09athuEe5kQfC7bV63jmGa,The People Tree,3OPuEGeKNwmxTfvOQJUh0Y,1993-01-01,13,Mother Earth,1v98YOkqO8vCNF8t2xUWiw,265760,Apple Green,0,...,7,-10.894,0,0.0651,0.005420,0.126000,0.1200,0.578,184.679,4
3,41s9iG3fTuu32yBnmDLCcZ,Mojo Hand,0X0Z6M1sQQpSRs9UbqpJEo,1962-04-13,9,Lightnin' Hopkins,6EZzVXM2uDRPmnHWq9yPDE,179733,Mojo Hand,0,...,11,-8.913,0,0.0640,0.000055,0.456000,0.3290,0.718,146.953,4
4,5F7gYvXfsmz6s98YOeZCvP,Poor and Infamous,08quDHgIgzN1sbHUhHsWfI,2020-09-25,12,Scumbag Millionaire,2OT0ERzAikorOYSIrRDjzA,205986,Inferno,0,...,6,-3.507,0,0.0537,0.000008,0.000008,0.3490,0.448,126.909,4


,index,track_id,genre_name
0,7,2sDurGdooRkMxYRGfRvPdn,Everything
1,14,00iwom2ONo0LvxvRkeXQdD,Everything
2,18,09athuEe5kQfC7bV63jmGa,Everything
3,24,41s9iG3fTuu32yBnmDLCcZ,Everything
4,31,5F7gYvXfsmz6s98YOeZCvP,Everything


We need to merge these two tables on `track_id` to get a single `DataFrame` with all the information we need. Since a track can have multiple genres (and thus can appear in multiple rows in the `TRAINING_DB_2` table), we'll need to aggregate the genres for each track so that each track only appears once in our merged `DataFrame`. If a track has multiple genres, they will be separated by commas in the `genre_name` column:

In [ ]:
genres_combined = (
    training_df.groupby("track_id")["genre_name"]
    .agg(lambda x: ",".join(sorted(set(x.str.strip()))))
    .reset_index()
)

genres_combined

,track_id,genre_name
0,000gn5FEkpxkwCx42ww99y,Indie Michoacano
1,000j3FwREid2RIpm6nTLn9,Korean City Pop
2,00189pkeAMRPFPeoLgYYp7,Armenian Folk
3,001OeQOJa8lLzu37AYBt2i,Vintage Classical Singing
4,001R5qTdlCLbDvuecoMoy3,Mindfulness
...,...,...
76437,7zycbRjPCH4NHGVjM5amfq,"Celtic Harp,Harp"
76438,7zymguaCu97mWiNH4d4G3T,Minneapolis Punk
76439,7zzNyY4AnIoW9zUSkCPnED,West Australian Hip Hop
76440,7zzaZggdFuadOs0R7SuJlp,"Hard Bop,Jazz Saxophone"


We perform an inner join on `track_id` to merge the two `DataFrame`s:


In [ ]:
merged_df = pd.merge(features_df, genres_combined, on="track_id", how="inner")

# For display purposes, we're showing columns that are semantic for humans:
merged_df[["track_id", "album_name", "artist_name", "genre_name"]]

,track_id,album_name,artist_name,genre_name
0,2sDurGdooRkMxYRGfRvPdn,Hifalutin,Lowly,"Aarhus Indie,Danish Indie Pop,Everything,Exper..."
1,00iwom2ONo0LvxvRkeXQdD,Accordéons de France,Karine Fontaine,"Accordéon,Everything"
2,09athuEe5kQfC7bV63jmGa,The People Tree,Mother Earth,"Acid Jazz,Everything"
3,41s9iG3fTuu32yBnmDLCcZ,Mojo Hand,Lightnin' Hopkins,"Acoustic Blues,Blues,Country Blues,Delta Blues..."
4,5F7gYvXfsmz6s98YOeZCvP,Poor and Infamous,Scumbag Millionaire,"Action Rock,Everything,Punk 'N' Roll,Swedish H..."
...,...,...,...,...
76437,3GpzybFTKLFrbGw6MYvaqQ,Xuân Hạ Thu Đông,Duong Trieu Vu,"Vietnam,Vietnam - Emerging"
76438,2cwxNwTxwTcvutGKpDBZmB,Dân Chơi Thôn,Tuấn Cry,"Vietnam,Vietnam - Emerging"
76439,3Gj1btfsYRTQ9icS0LfvAx,For You,Scenery,"Vietnam,Vietnam - Underground"
76440,5RBPhmnbILMVrHdl8pD2zp,Họ Gọi Anh Là Máy Bay Lag3,DJ Tom2k,"Vietnam,Vietnam - Underground"


Following this, we take a 10% random sample of the merged `DataFrame`:


In [ ]:
np.random.seed(42)

sampled_df = merged_df.sample(frac=0.1)

# Take a look at the indices of the sampled DataFrame
sampled_df[["track_id", "album_name", "artist_name", "genre_name"]]

,track_id,album_name,artist_name,genre_name
37147,2anLSOW0qJmimTA9jWBq90,Robert Johnson,Fínix MG,"Hip Hop Tuga,Trap Tuga"
38282,6TyXeawZxBzMUq193xscwA,Kala Masala,Sahir,Indian Lo-Fi
57449,6dDBOwTKKP1VwLrlmC2UJa,Timeless Love,Love Hits On Piano,Piano Cover
53282,1Qpdc3wbQfN44u8Nl0LNPm,The Answer To Your Problems?,Marky Ramone & the Intruders,Neo-Proto
3620,7EtQeUXeYIhMBbV7uO7SKQ,Manic,Knockout Kid,"Anthem Emo,Easycore"
...,...,...,...,...
25916,17mcxzjv3w5Oz7fXXfJxxv,Dixieland Hits,Louisiana Blues Band,Dixieland
5986,1K6PKHxFILfVQfLk7ac6wa,Sister Phantom Owl Fish,Trevor Dunn's Trio Convulsant,"Avant-Rock,Laboratorio"
1550,22uAbeSGb5dh2wh3c6Akkw,Kiekies,Lucas Maree,Afrikaans
45919,6BU4THl6Q6JPSjDIpOjDqm,De NG La Banda Para Curaçao,NG La Banda,"Latin Jazz,Salsa,Timba"


For some reason there's a bunch of genres called "Everything" that are polluting the genres column. If it is one of multiple genres for a song, we'll remove it from the cell. If it is the only genre for a song, we'll remove the song from the dataset:

In [ ]:
sampled_df["genre_name"] = sampled_df["genre_name"].str.replace("Everything", "")
sampled_df = sampled_df[sampled_df["genre_name"] != ""]

Finally, the most tedious part of this section is creating an empty JSON file and manually filling in genre-to-country associations. We followed a strict methodology:
- We only used countries from the `pycountry` library (which uses [ISO 3166-1](https://en.wikipedia.org/wiki/ISO_3166-1) country names)
- We only mapped genres that are associated with a particular country today. For genres that were developed in a particular country but have since expanded globally (e.g. hip-hop, reggae), we did not map them to that country.
- We excluded genres that are associated with ethnic groups that don't majority associate to an existing country, such as the Kurdish and Hmong people.

This manual mapping process helps us analyze geographic patterns in music genres while maintaining data consistency. This was the script that we used to create the JSON file:

In [ ]:
unique_genres = set()
for genres in sampled_df["genre_name"].str.split(","):
    if genres is not None:
        unique_genres.update([genre.strip() for genre in genres])

genre_country_map = {genre: "" for genre in sorted(unique_genres)}

with open("data/genre_country_mapping.json", "w", encoding="utf-8") as f:
    json.dump(genre_country_map, f, indent=4, ensure_ascii=False)

This creates a JSON file that looks like such:

```json
{
    "": "",
    "2020–1960": "",
    "21st Century Classical": "",
    "432Hz": "",
    "48G": "",
    "A Cappella": "",
    "A3": "",
    "ACTORS": "",
    "ASMR": "",
    "ATL Hip Hop": "",
    "ATL Trap": "",
    "Aarhus Indie": "",
    "Aberdeen Indie": "",
    "Abstract": "",
    // ...
}
```

Here is an example segment of the JSON file that we created, `data/filled_genre_country_mapping.json`:

```json
{
    // ...
    "Azeri Pop": "Azerbaijan",
    "Azeri Rap": "Azerbaijan",
    "Azeri Traditional": "Azerbaijan",
    "Azonto": "Ghana",
    "Azontobeats": "Ghana",
    "BC Underground Hip Hop": "Canada",
    "Bachata": "Dominican Republic",
    "Bachata Dominicana": "Dominican Republic",
    "Background Jazz": "",
    "Background Music": "",
    // ...
}
```

Adding the `country` column to our dataset:

In [ ]:
with open('data/filled_genre_country_mapping.json', 'r', encoding='utf-8') as f:
    genre_country_map = json.load(f)

def get_country_from_genres(genre_str):
    genres = genre_str.split(',')
    countries = []
    for genre in genres:
        if genre in genre_country_map and genre_country_map[genre]:
            countries.append(genre_country_map[genre])

    if not countries:
        return ''

    unique_countries = list(set(countries))

    # If more than 4 different countries, return empty string
    if len(unique_countries) > 4:
        return ''

    # If multiple countries, return most common one
    if len(unique_countries) > 1:
        return max(countries, key=countries.count)

    return countries[0]

sampled_df['country'] = sampled_df['genre_name'].apply(get_country_from_genres)

sampled_df[['album_name', 'artist_name', 'genre_name', 'country']]


,album_name,artist_name,genre_name,country
37147,Robert Johnson,Fínix MG,"Hip Hop Tuga,Trap Tuga",Portugal
38282,Kala Masala,Sahir,Indian Lo-Fi,India
57449,Timeless Love,Love Hits On Piano,Piano Cover,
53282,The Answer To Your Problems?,Marky Ramone & the Intruders,Neo-Proto,
3620,Manic,Knockout Kid,"Anthem Emo,Easycore",
...,...,...,...,...
25916,Dixieland Hits,Louisiana Blues Band,Dixieland,
5986,Sister Phantom Owl Fish,Trevor Dunn's Trio Convulsant,"Avant-Rock,Laboratorio",
1550,Kiekies,Lucas Maree,Afrikaans,South Africa
45919,De NG La Banda Para Curaçao,NG La Banda,"Latin Jazz,Salsa,Timba",Cuba


Saving the sampled dataset to a `.csv` file:

In [ ]:
sampled_df.to_csv("data/last_fm_sample.csv", index=False)

Our data is now ready for analysis! Recall that we were working on a 10% random sample of the 10% random sample of the full dataset, so here is our actual `.csv` file (simply a 10% random sample of the full dataset) that we will use for analysis:

In [ ]:
data = pd.read_csv("data/last_fm.csv")
data[['album_name', 'artist_name', 'genre_name', 'country']]

,album_name,artist_name,genre_name,country
0,Søndermarken (Video Edit),Henrik Lindstrand,Neo-Classical,NaN
1,For You (LP),Ago,Deep Dubstep,NaN
2,Astronauta Tupy,Pedro Luis E A Parede,"Brazilian Indie,Manguebeat",Brazil
3,The Garbageman and The Prostitute,Kill Me Tomorrow,Experimental Indie Rock,NaN
4,Evergreen,Bin Juice,Sydney Indie,Australia
...,...,...,...,...
76437,Swan Song,Neil Mukherjee,Indian Instrumental Rock,India
76438,Opéra Puccino (Edition Collector),Oxmo Puccino,"French Hip Hop,Old School Rap Français,Rap Con...",France
76439,Refractal - EP,Refractal,Baton Rouge Indie,United States
76440,Cayenne,Lunar,Trap Funk,Brazil


# Results

## Exploratory Data Analysis

We first start with a univariate, surface-level analysis of the data we are provided—this helps us contextualize how the dataset quantifies audio characteristics. We extract key columns into a hardcoded list `audio_features` containing numerical data, and then create a large subplot showing the distribution and mean of each column:

In [ ]:
audio_features = [
    "danceability",
    "energy",
    "loudness",
    "speechiness",
    "instrumentalness",
    "acousticness",
    "liveness",
    "valence",
    "tempo",
    "track_popularity",
]

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76442 entries, 0 to 76441
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   track_id            76442 non-null  object 
 1   album_name          76441 non-null  object 
 2   album_id            76442 non-null  object 
 3   album_release_date  76442 non-null  object 
 4   album_total_tracks  76442 non-null  int64  
 5   artist_name         76439 non-null  object 
 6   artist_id           76442 non-null  object 
 7   track_duration      76442 non-null  int64  
 8   track_name          76440 non-null  object 
 9   explicit            76442 non-null  int64  
 10  track_popularity    76442 non-null  int64  
 11  id                  76442 non-null  object 
 12  danceability        76442 non-null  float64
 13  energy              76442 non-null  float64
 14  key                 76442 non-null  int64  
 15  loudness            76442 non-null  float64
 16  mode

In [ ]:
fig = make_subplots(
    rows=3,
    cols=4,
    subplot_titles=[
        f"{feature.replace('_', ' ').title()}\nMean: {data[feature].mean():.2f}"
        for feature in audio_features
    ],
)
fig.update_layout(
    height=800,
    width=1200,
    title_text="Distribution of Audio Features",
    showlegend=False,
)

for idx, feature in enumerate(audio_features):
    row = idx // 4 + 1
    col = idx % 4 + 1

    mean_val = data[feature].mean()

    fig.add_trace(
        go.Histogram(x=data[feature], name=feature, nbinsx=30), row=row, col=col
    )

    fig.add_vline(x=mean_val, line_dash="dash", line_color="red", row=row, col=col)

fig.update_xaxes(tickangle=45)
fig.update_layout(showlegend=False)

fig.show()

This subplot helps display what amount of each characteristic is common within our dataset.

- **Danceability** indicates that most songs fall within the middle and create a mostly-normal distribution to highlight that most songs are semi-danceable and the extremes are either not danceable at all or completely danceable.

- **Energy** of songs features a mostly linear correlation between the count of songs and their energy, with songs leaning more towards higher energy. This is in line with the idea that our songs are randomly sampled and should have a mean somewhere in the middle values to represent a diverse data set.

- The **loudness** graph indicates that most songs hover around a certain loudness and skew left from the peak, showcasing how most music is mastered at a certain volume and stays within a certain loudness. The quieter songs are either generally softer in totality or just under mastered to a certain loudness.

- Our **speechiness** data indicates that most of our songs feature very little to none spoken word and are mostly singing and instrumental. It also shows that if songs do have spoken word, it isn’t any more than 40% of the song, meaning we don’t have any spoken word pieces.

- **Acousticness** highlights how songs can either be fully acoustic or fully non-acoustic with most songs being a mix of both. This makes sense because there are songs with only acoustic instruments such as folk and songs with no acoustic instruments such as EDM. Most songs fall more towards less acoustic-ness because of the larger amount of virtual and electronic instruments that exist and their wider genre appeal.

- The **liveness** graph shows that most songs have a low liveness value which correlates with how most songs are recorded in a studio with a less-live setup. Since every song is performed at least a little bit live (recording takes), the graph has very little 0 counts but averages around the 0.1 mark.

- **Valence** indicates how happy/sad our songs are with there being a clear range from “sad” to “happy” songs. We can see that while the average song is somewhere between the two valences, there are also plenty of songs that are completely extreme with either a totally 0 or totally 1 valence.

- Our **tempo** data makes sense as we see songs fall within 80 to 160 bpm with some outliers. Those outliers are typically from very mellow genres such as ambient music and extreme genres like thrash metal.

- **Track popularity** illustrates how most songs are not popular at all with a skew right indicating how less songs are popular than not. This makes sense as we expect randomly sorted songs to include some that are extremely unknown and barely any that are famous. The massive dropoff from 0 to the next integer shows how rare known music is in actuality.


In [ ]:
correlation_matrix = data[audio_features].corr()

corr_df = correlation_matrix.reset_index().melt(id_vars='index')
corr_df.columns = ['x', 'y', 'correlation']

fig = px.imshow(
    correlation_matrix,
    labels=dict(x="Features", y="Features", color="Correlation"),
    x=correlation_matrix.columns,
    y=correlation_matrix.columns,
    color_continuous_scale='RdBu',
    aspect='equal'
)

# Add text annotations manually
for i in range(len(correlation_matrix.columns)):
    for j in range(len(correlation_matrix.columns)):
        fig.add_annotation(
            x=correlation_matrix.columns[i],
            y=correlation_matrix.columns[j],
            text=str(round(correlation_matrix.iloc[j, i], 2)),
            showarrow=False,
            font=dict(color="black")
        )

fig.update_layout(
    title='Correlation Matrix of Audio Features',
    title_x=0.5,
    width=800,
    height=800
)

fig.show()

print("\nStrongest Correlations:")
print("-----------------------")
correlations = []
for i in range(len(audio_features)):
    for j in range(i + 1, len(audio_features)):
        corr = correlation_matrix.iloc[i, j]
        if abs(corr) > 0.3:
            correlations.append((audio_features[i], audio_features[j], corr))

for feat1, feat2, corr in sorted(correlations, key=lambda x: abs(x[2]), reverse=True):
    print(f"{feat1} vs. {feat2}: {corr:.3f}")


Strongest Correlations:
-----------------------
energy vs. loudness: 0.789
energy vs. acousticness: -0.748
loudness vs. acousticness: -0.608
danceability vs. valence: 0.516
loudness vs. instrumentalness: -0.366
instrumentalness vs. valence: -0.313
loudness vs. valence: 0.310


The heatmap above allows us to visualize the correlation matrix between multiple audio characteristics using colors and a pairwise correlation coefficient, $r$, which can range from -1.00 to +1.00. Positive correlations are highlighted in warm tones, and negative relationships are highlighted in cool tones. For our project, correlations exceeding the absolute coefficient value of 0.3 are considered a "strong correlation." A more red square and a higher positive coefficient value indicates a stronger positive correlation between two audio characteristics. A more blue square and a lower negative coefficient value indicates a stronger negative correlationship between two audio characteristics. A more grey square and coefficient value of 0.00 means no correlation between two audio characteristics.

As expected, diagonal audio characteristics have a correlation coefficient of +1.00 (They have a perfect correlation with themselves).

The strongest positive correlation between two audio characteristics is energy vs. loudness, which has a correlation coefficient of +0.789. This means:
- As the energy of a song increases, its loudness tends to also increase.

On the other hand, the strongest negative correlation between two audio characteristics is energy vs. acousticness, with a correlation coefficient of -0.748. This means:
- As the energy of a song increases, its acousticness tends to decrease.

There is moderately strong positive correlation between danceability vs. valence with a correlation coefficient of 0.516 and between loudness vs. valence with a correlation with a correlation of 0.310. This means:
- As the danceability of a song increases, its valence tends to increase.
- As the loudness of a song increases, its valence also tends to increase.

There is a moderately strong negative correlation between loudness vs. acousticness with a correlation coefficient of -0.608, between loudness vs. instrumentalness with a correlation coefficient of -0.366, and between instrumentalness vs. valence with a correlation coefficient of -0.313. This means:
- As the loudness of a song increases, its acousticness tends to decrease.
- As the loudness of a song increases, its instrumentalness tends to decrease.
- As the instrumentalness of a song increases, its valence tends to decrease.

Audio characteristics with near-zero correlation coefficients (e.g. speechiness and tempo) have an extremely weak linear relationship.

In [ ]:
genre_counts = (
    data['genre_name']
    .str.split(',')
    .explode()
    .value_counts()
    .head(20)
)

fig = px.bar(
    x=genre_counts.index,
    y=genre_counts.values,
    title='Top 20 Most Common Music Genres',
    labels={'x': 'Genre', 'y': 'Number of Tracks'},
    color_discrete_sequence=['rgb(67,147,195)']
)

fig.update_layout(
    title_x=0.5,
    title_font_size=20,
    xaxis_tickangle=45,
    showlegend=False,
    height=600,
    width=650
)

fig.update_traces(
    text=genre_counts.values,
    textposition='outside'
)

fig.show()

The above chart displays the top 20 most common genres within the dataset. Something very notable about this chart is the recurring amount of genres that make use of electronic music and instruments, specifically **'dance pop'**, **'EDM'**, **'pop dance'**, **'electropop'**, and **'tropical house'**. Furthermore, the ratio of electronic genres seems to be more common compared to other umbrella genres like rock and hip-hop.

* This info further supports the distribution graphs above, specifically the graphs about 'acousticness' and 'danceability'. As noted earlier, most songs fall towards less use of acoustics given the larger variety of electronic and virtual instruments, and, most relevant, "wider genre appeal".

* As for the distribution chart about 'danceability', whereas the chart seems to find that most songs fall in between extremely non-danceable and extremely danceable, the slope leans closer to the danceable extreme. This too is reflected in the 20 most common genres above not just in the recurring amounts of genres under the electronic umbrella, but also other notably upbeat genres such as **'pop'**, **'K-pop'** and **'Latin'**. Of course, it could also be argued that this overall balance between non-danceable and very danceable is also reflected in how other common genres are very much also ones that aren't as known to be danceable, such as **'hip-hop'**, **'classical performance'** and **'rock'**.

* The distribution graph about 'energy' can also be connected to our top 20 chart. As stated earlier, the original chart about energy has an overall positive linear correlation between the amount of songs and energy. While there are significant outliers such as **'classical performance'** and **'indie folk'**, nearly every genre on this list is some form of energetic.

* 'Valence', or rather the balance between overall happy and sad moods, is also reflected in this whole list of genres. In general, most of these genres like **'pop'**, **'rock'**, and **'hip-hop'** have their fair share of happy and sad songs. That said, notice also how genres that have more of a reputation for aggressive or sad moods such as metal or ambience aren't within the list, even despite how both often fall under the 'rock' and 'electronic' umbrellas respecitvely.

* The exact same observations could also be applied to how our genre chart connects to our 'tempo' distribution chart. In general, there seems to be a preference for songs with the average tempo noted above. Again, notice how whereas there is a wide preference for very energetic, pretty danceable songs, genres on both extremes of tempo like, once again, metal and ambience, don't appear on the genre chart. Thus, this too reflects the overall preference for tempos between 80 bpm and 160 bpm, any and all of which can be found in **'electronic'**, **'rock'**, **'pop'** and **'hip-hop'** genres easily.

We can also get a good look at the distribution of genres by country, using a logarithmic scaling to better deal with outliers like the United States and Brazil (which have significantly more genres than other countries):

In [ ]:
with open('data/filled_genre_country_mapping.json', 'r', encoding='utf-8') as f:
    genre_country_map = json.load(f)

country_counts = {}
for genre, country in genre_country_map.items():
    if country:
        country_counts[country] = country_counts.get(country, 0) + 1

country_data = pd.DataFrame.from_dict(country_counts, orient='index', columns=['genre_count'])
country_data.index.name = 'country'
country_data = country_data.reset_index()

country_data['log_genre_count'] = np.log10(country_data['genre_count'])

fig = px.choropleth(
    country_data,
    locations='country',
    locationmode='country names',
    color='log_genre_count',
    hover_name='country',
    hover_data={'genre_count': True, 'log_genre_count': False},
    color_continuous_scale='Viridis',
    title='Global Distribution of Music Genres by Country (Log Scale)',
    labels={'log_genre_count': 'Number of Genres (log10)', 'genre_count': 'Number of Genres'}
)

fig.update_layout(
    title={
        'font_size': 24,
        'x': 0.5,
        'xanchor': 'center',
        'y': 0.95,
        'yanchor': 'top'
    },
    geo=dict(
        showframe=False,
        showcoastlines=True,
        projection_type='equirectangular',
        landcolor='rgb(243, 243, 243)',
        oceancolor='rgb(204, 229, 255)',
        showcountries=True,
        countrycolor='rgb(204, 204, 204)'
    ),
    width=1200,
    height=700,
    margin=dict(l=40, r=40, t=60, b=40),
    paper_bgcolor='white',
    plot_bgcolor='white'
)

fig.show()

The choropleth map above provides insights into how music genres are distributed across different countries in our dataset. The United States emerges as the clear leader in genre diversity, followed by Brazil and major European nations like the United Kingdom, France, and Germany. This pattern likely reflects both the size of these music markets and Spotify's stronger presence in these regions. We can observe clear regional patterns, with Western countries generally showing greater genre diversity, while Latin American countries demonstrate moderate genre counts. Many African and Asian countries show fewer documented genres, though this may reflect data limitations rather than actual musical diversity.

The visualization also reveals potential sampling biases in our dataset. Some countries appear empty due to missing data, and the genre counts may be more indicative of Spotify's market availability than actual musical diversity. Additionally, Western music classification systems may not adequately capture genre distinctions in non-Western musical traditions. Despite these limitations, we can identify several regional cultural hubs with notably higher genre counts, such as Brazil in South America, Japan and China in East Asia, and South Africa in Africa.

This geographic distribution of genres helps us understand the global musical landscape while reminding us to consider both data limitations and cultural context in our analysis. The logarithmic scale used in the visualization helps to better highlight differences between countries, as the raw counts vary dramatically between the most and least represented nations.

## Analysis of Audio Features Across Top 10 Countries

Next, we'll analyze how various audio features like danceability, energy, tempo, etc. vary across the top 10 countries with the most tracks in our dataset (we balance the sample size for each country to ensure a fair comparison). This will help us understand if there are distinct musical characteristics associated with different regions. We'll visualize these patterns using a series of bar charts comparing the average values of each audio feature.

In [ ]:
audio_features = [
    "danceability",
    "energy",
    "loudness",
    "speechiness",
    "instrumentalness",
    "acousticness",
    "valence",
    "tempo",
]

data_filtered = data[data['country'] != '']
top_10_countries = data_filtered['country'].value_counts().head(10)
min_tracks = top_10_countries.min()

balanced_data = []
for country in top_10_countries.index:
    country_data = data_filtered[data_filtered['country'] == country]
    sampled_data = country_data.sample(n=min_tracks, random_state=42)
    balanced_data.append(sampled_data)

data_filtered = pd.concat(balanced_data)
country_means = data_filtered.groupby('country')[audio_features].mean().reset_index()

n_features = len(audio_features)
n_cols = 2
n_rows = (n_features + 1) // 2

fig = make_subplots(
    rows=n_rows,
    cols=n_cols,
    subplot_titles=[f'Average {feature.title()} by Country' for feature in audio_features]
)

for i, feature in enumerate(audio_features):
    row = i // n_cols + 1
    col = i % n_cols + 1

    feature_data = country_means.sort_values(feature, ascending=False)

    fig.add_trace(
        go.Bar(
            x=feature_data['country'],
            y=feature_data[feature],
            name=feature,
            marker=dict(
                color=feature_data[feature],
                colorscale='Viridis'
            )
        ),
        row=row,
        col=col
    )

    fig.update_xaxes(tickangle=45, row=row, col=col)
    fig.update_yaxes(title_text=feature.title(), row=row, col=col)

fig.update_layout(
    height=300*n_rows,
    width=1200,
    showlegend=False,
    title_text="Audio Features by Country (Balanced Sample)",
    title_x=0.5
)

fig.show()

For danceability, we see that Mexico has the highest score but is also closely followed by Brazil, with a visible gap between the top two and the next highest in Germany. This may be due to the culture of Mexico and Brazil incorporating a large amount of dancing and therefore popularizing music that relates to it.

Mexico again has the loudest music, but this time being closely followed by Japan and again Brazil. We see though that the top 10 countries are very close in loudness with only a relative difference of about 3 dB between them all. We also see that countries in Western Europe and those with traditionally Western culture (The U.S. and Canada) are grouped together. It altogether shows that music doesn't have that much difference in terms of loudness within these country's samples.Furthermore, the loudness level may be due to the difference in techniques of mixing and mastering across the genres or locations.


With instrumentalness, the score seems to swap with Mexico and Brazil now the lowest in the group. We can see that overall, instrumentalness of songs is relatively low with only the highest average being 0.3 (30%). There is also a drop-off between Japan and Spain showing how its only the top five countries that have a significant amount of instrumentalness.

Valence is mostly 0.5 across the board showing how music is split evenly between being happy and sad for most countries. Mexico is the outlier here with a valence above 0.6 indicating that the average song is happier than it is sad. This may coincide with its danceability and loudness reflecting a non-melancholy and more uplifting sound.

In lieu of that, Mexico again has a high average energy to its songs with other countries having an energy score matching that of their danceability such as Brazil and Spain. However, the average energy across the countries is very similar with there being at most a 0.1 amount of difference between them all. Similar to valence, this is likely due to the 50/50 amount of sad and happy songs or many songs not leaning too hard into a certain binary emotion.

Most songs have very little speechiness as shown with the highest score being an outlier of around 0.15 in Germany, showing that it may has some cultural interest in spoken word songs. The other scores illustrate how most songs have very little, if any, spoken word in their music.

Acousticness has averages below 0.5 to show that most songs are not mostly acoustic. We see the France also has the lowest energy, tempo, and highest instrumentalness suggesting a correlation between those factors. In general, most songs have around a 0.3 level of acousticness indicating how acoustic instruments exist in music but may not be as popular as electronic instruments.

Finally, the tempo graph shows how normalized BPM is across countries, with their averages all hovering around 120-130 BPM. It seems to be largely unrelated to other factors or geographical location. However, it should be noted that the countries displayed are more economically stable than much of the world and may represent a different style of music that is dominated by popular culture rather than more traditional or "underground" music.



## Analyzing Genres Across Different Countries

### Hip-Hop/Rap

Below, we identify countries with significant hip-hop/rap presence in our dataset and analyze how different countries put their own spin on the genre through audio features. This allows us to explore how hip-hop/rap has been adapted and interpreted across different cultures:

In [ ]:
hiphop_mask = data["genre_name"].str.lower().str.contains("hip hop|rap", na=False)
hiphop_data = data[hiphop_mask].copy()

counts = hiphop_data["country"].value_counts()
selected_countries = counts[counts >= 200].index.tolist()

print("\nCountries with at least 200 hip-hop/rap tracks:")
display(selected_countries)


Countries with at least 200 hip-hop/rap tracks:


['United States', 'Brazil', 'Germany', 'Canada', 'France']

We select the five countries that each has at least 200 hip-hop/rap tracks to analyze: The United States (where hip-hop was born), Canada, Brazil, France, and Germany. These countries were chosen due to their geographical and significant cultural differences from each other, which we expect to be reflected in their audio features. To statistically validate any differences between these countries' hip-hop styles, we perform the Kruskal-Wallis H-test, which is a non-parametric test that determines whether samples from two or more groups come from the same distribution. The test computes an $H$-statistic and corresponding $p$-value. The $H$-statistic measures the extent of differences between the mean ranks of the groups—a larger $H$-statistic suggests greater differences between groups. When combined with the $p$-value (where $p < 0.05$ is considered significant), we can determine if the observed differences between countries' hip-hop styles are statistically meaningful or likely due to chance:

In [ ]:
def analyze_genre_by_country(
    data,
    countries,
    genre_regex,
    sample_size=200,
    chart_title="Audio Feature Distributions by Country",
):
    country_mask = data["country"].isin(countries)
    genre_mask = data["genre_name"].str.lower().str.contains(genre_regex, na=False)
    filtered_data = data[country_mask & genre_mask].copy()

    sampled_data = []
    for country in countries:
        country_data = filtered_data[filtered_data["country"] == country]
        if len(country_data) >= sample_size:
            sampled = country_data.sample(n=sample_size, random_state=42)
            sampled_data.append(sampled)
        else:
            print(
                f"Warning: {country} has fewer than {sample_size} tracks ({len(country_data)} tracks)"
            )
            sampled_data.append(country_data)

    filtered_data = pd.concat(sampled_data)

    fig = make_subplots(
        rows=3, cols=3, subplot_titles=[f.title() for f in audio_features]
    )

    for i, feature in enumerate(audio_features):
        row = i // 3 + 1
        col = i % 3 + 1

        box_trace = go.Box(
            x=filtered_data["country"],
            y=filtered_data[feature],
            name=feature.title(),
            showlegend=False,
        )

        fig.add_trace(box_trace, row=row, col=col)
        fig.update_xaxes(row=row, col=col)
        fig.update_yaxes(title_text=feature.title(), row=row, col=col)

    fig.update_layout(height=900, width=1200, title_text=chart_title, showlegend=False)

    fig.show()

    feature_names = []
    h_stats = []
    p_values = []
    significances = []

    for feature in audio_features:
        country_data = [
            filtered_data.loc[filtered_data["country"] == country, feature].dropna()
            for country in countries
        ]

        stat, p = kruskal(*country_data)

        feature_names.append(feature.title())
        h_stats.append(round(stat, 4))
        p_values.append(round(p, 4))
        significances.append("Significant" if p < 0.05 else "Insignificant")

    kw_results = pd.DataFrame(
        {
            "Feature": feature_names,
            "H-statistic": h_stats,
            "p-value": p_values,
            "Significance": significances,
        }
    ).set_index("Feature")

    display(kw_results)

In [ ]:
analyze_genre_by_country(
    data,
    selected_countries,
    "hip hop|rap",
    chart_title="Hip Hop/Rap Music Audio Feature Distributions by Country"
)

,H-statistic,p-value,Significance
Feature,,,
Danceability,5.9647,0.2018,Insignificant
Energy,31.9039,0.0000,Significant
Loudness,25.0444,0.0000,Significant
Speechiness,14.2436,0.0066,Significant
Instrumentalness,14.5264,0.0058,Significant
Acousticness,69.5881,0.0000,Significant
Valence,18.9734,0.0008,Significant
Tempo,5.9087,0.2061,Insignificant


The bar plot graph above allows us to visualize the distributions of different audio characteristics in hip-hop tracks across our five chosen countries (the United States, Canada, Brazil, France, and Germany). Each color-coded subplot summarizes the range and median values of a certain audio characteristic (e.g tempo, dancebility, energy) for hip-hop tracks in each of the countries.

From the bar plot, we see the following trends for each audio feature:
* **Danceability:** The median dancebility is relatively similar among the five countries. However, Brazil has slightly more variability based on its interquartile range. The United States and Canada have smaller IQRs, so we can infer there is higher consistency in the danceability of hip-hop tracks from those countries.
* **Energy:** The median energy is similar among the five countries. However, Brazil has slightly greater range, and we see that some Brazilian tracks have higher energy levels. This could reflect the possibility of local influence on the hip-hop genre. We also note that Germany has one very low energy outlier which could be due to experimental style, genre fusion, or local influence.
* **Loudness:** The median loudness is similar among all countries, but Germany has the highest loudness median level while Brazil has the lowest loudness median level. Germany has the most consistent level of loudness with the smallest IQR while Brazil has more variability in loudness. Outliers with lower loudness levels are present in all five countries, with Brazil having the most significant outlier at a loudness level of -20.157.
* **Speechiness:** The United States appears to have the highest speechiness level median at 0.237 while Canada and Brazil have the lowest spechiness medians at 0.1685 and 0.164 respectively. The United States might have the highest median speechiness as it is the birthplace of hip-hop, and many American hip-hop artists place emphasis on lyricism, storytelling, and flow, or the manner in which an artist delivers their lyrics, naturally elevating speechiness levels. On the other hand, Canadian and Brazilian hip-hop may have lower median speechiness levels due to more melodic influence from regional styles of music. All five countries share similar variability in speechiness levels with their IQRs overlapping one another.
* **Instrumentalness:**  Tracks from all five countries have very low values of instrumentalness, with most close to 0, reflecting the vocal-heavy nature of hip-hop across all cultures. However, a few outliers in Brazil and Canada suggest some experimentation with more instrumental tracks.
* **Acousticness:** Acousticness levels vary drastically across the five countries. Brazil and France have the highest acoustic medians at 0.26 and 0.2455 respectively. This could possibly due to regional musical influence on the hip-hop genre. The United States has the lowest acoustic median at 0.1135. The United States has the lowest variability in acousticness levels while France has the highest variability.
* **Valence:** Valence levels are mostly similar across the five countries. France has the highest valence median of 0.5715 while Canada has the lowest valence median of 0.482.
* **Tempo:** Brazil has the highest median tempo of 123 beats per minute (BPM) while the United States and Germany have the lowest median tempos of 109.282 BPM and 108.1175 BPM respectively. The United States has the greatest variability in tempo while Brazil has more consistency in tempo within the hip-hop genre.

Furthermore, based on our results of the Kruskal-Wallis H-test for each audio characteristic, we see there is a **significant difference in the energy, loudness, speechiness, instrumentalness, acousticness, and valence levels** among the five countries as their p-values are all less than 0.05. We also see that there is **NO significant difference in the danceability and tempo levels across the five countries** as their p-values are more than or equal to 0.05.

### Rock

We can do something similar for rock:

In [ ]:
analyze_genre_by_country(
    data,
    ["United States", "United Kingdom", "Germany", "Sweden", "Japan"],
    "rock",
    chart_title="Rock Music Audio Feature Distributions by Country"
)

,H-statistic,p-value,Significance
Feature,,,
Danceability,15.3965,0.0039,Significant
Energy,93.8082,0.0000,Significant
Loudness,167.6329,0.0000,Significant
Speechiness,22.6880,0.0001,Significant
Instrumentalness,48.2653,0.0000,Significant
Acousticness,53.7056,0.0000,Significant
Valence,18.9025,0.0008,Significant
Tempo,7.7425,0.1015,Insignificant


* **Danceability:** All countries overall are in the middle range for danceability. There is an outlier in the United States that is very high in danceability. Likewise, there are outliers in Japan and Germany that are very low on the scale.
* **Energy:** Energy ranks high on all five countries, which is to be expected of rock music. Japan significantly ranks higher in energy compared to the other four countries. That said, Japan still has outliers that are on the opposite. In comparison, Germany has a wider average range of energy in their rock, seeing how the upper fence and lower fence equal the max and min at 0.992 and 0.0106 respectively. Thus, it's worth noting Germany covers pretty much the entire spectrum.
* **Loudness:** All countries overall scale high on the loudness scale, which like energy is also to be expected of rock as a whole. However, Japan has a significantly higher level of loudness compared to the other countries on this list with a median of -4.326. Germany also has a significant number of outliers that rank lower on the loudness scale, the lowest in fact being -31.748.
* **Speechiness:** In general, all five countries lean lower on the spectrum of speechiness. This makes fair sense given how rock doesn't have as much emphasis on vocals over instrumentals compared to say, hip-hop as described above. However, there are still multiple outliers in all five countries that rank significantly higher in speechiness, especially in the United States.
* **Instrumentalness:** In turn, there's significantly higher rates of instrumentality in the United States, the United Kingdom, and Germany. However, Sweden and Japan actually have far lower average rates of instrumentality, especially in proportion to how they ranked in speechiness to the point where their respective data is composed of mostly outliers.
* **Acousticness:** Interestingly, all countries rank lower on the acousticness spectrum, incidating that while acoustic instruments are still used, electronic instruments are more frequently used in tandem. Granted, this could be a matter of common rock instruments like electric guitars and synthesizers being counted as 'non-acoustic'. The United States in particular seems have the widest quartile range in acousticness, while Japan seems pretty firmly in the more electronic end, though there's still multiple outliers towards the fully acoustic end as well.
* **Valence:** All five countries are pretty similar in terms of valence in their rock music, although Sweden leans closer to bottom at 0.42 whereas Japan leans closer to the top at 0.569.
* **Tempo:** Tempo also seems pretty even between all five countries, with the United Kingdom having the highest median at 130.0185. Worth noting is Japan has a higher q3 at 163.9485

### Pop

In [ ]:
analyze_genre_by_country(
    data,
    ["United States", "United Kingdom", "China", "Japan", "Korea, Republic of"],
    "pop",
    chart_title="Pop Music Audio Feature Distributions by Country"
)

,H-statistic,p-value,Significance
Feature,,,
Danceability,30.9517,0.000,Significant
Energy,152.9365,0.000,Significant
Loudness,196.9186,0.000,Significant
Speechiness,50.8179,0.000,Significant
Instrumentalness,111.5183,0.000,Significant
Acousticness,136.3638,0.000,Significant
Valence,81.8813,0.000,Significant
Tempo,8.6328,0.071,Insignificant


From the bar plot, we see the following trends for each audio feature:
* **Danceability:** The median danceability is relatively similar among the five countries. South Korea has the highest median danceability of 0.644 while the United Kingdom has the lowest median danceability of 0.547. The variability in danceability is also similar across the five countries but the United States has the highest variability based on its IQR.
* **Energy:** The median energy varies greatly among the five countries. Japan has the highest median energy level at 0.81, while China has the lowest median energy level at 0.4825. The United Kingdom has the greatest variability in energy while Japan has the least variability.
* **Loudness:** The median loudness is the greatest in Japan and South Korea with levels of -4.906 and -4.678 respectively. The median loudness is the least in China with a level of -8.172. Variability in loudness appears to be the least in Japan and South Korea, and the greatest in the United States and the United Kingdom. While all five countries have lower loudness outliers, South Korea has a significant outlier with a loudness level of -25.745.
* **Speechiness:** The median speechiness is similar across all five countries with low levels close to 0. Besides China, all countries have a median speechines close to 0.04. The United States has the greatest variability in speechiness levels while China has the least variability. All five countries have high speechiness outliers. The U.S has the highest speechiness outlier at a level of 0.372.
* **Instrumentalness:**The median instrumentalness is similar across all five countries with low levels close to 0. The United Kingdom appears to have the greatest variability in instrumentalness. All five countries have high intstrumentalness outliers, with the highest outlier being from South Korea with an instrumentalness level of 0.968.
* **Acousticness:** Acousticness levels vary drastically across the five countries. China has the highest median acousticness level of 0.497 while the United Kingdom has the lowest median acousticness level of 0.00604. South Korea has the greatest variability in acousticness with a wider IQR while Japan has the least variability with a tighter IQR.
* **Valence:** Valence levels are mostly similar across the five countries, with the exception of China. Japan has the highest median valence at 0.5815 while China has the lowest median valence at 0.343. The United Kingdom has the greatest variability in valence levels while China and Japan appear to have the lowest variability in valence levels. There is a single higher outlier from China with a valence level of 0.973.
* **Tempo:** Tempo is relatively similar across the five countries. Japan has the highest median tempo of 123.989 beats per minute (BPM) while the United States has the lowest median tempo of 114.2925 BPM. Japan has the highest variability in tempo while South Korea has the lowest variability in tempo.


Furthermore, based on our results of the Kruskal-Wallis H-test for each audio characteristic, we see there is a **significant difference in the danceability, energy, loudness, speechiness, instrumentalness, acousticness, and valence levels** among the five countries as their p-values are all less than 0.05. We also see that there is only **NO significant difference in the tempo across the five countries** as their p-values are more than or equal to 0.05.

# Ethics & Privacy


Our analysis of music characteristics across cultural boundaries presents several important ethical considerations that we must address before and throughout the research process.

While Spotify's API provides rich data, we acknowledge that there are potential issues with particular data collection methods and usage rights. We recognize that using Spotify as our primary data source introduces inherent sampling bias, as Spotify’s availability and adoption varies significantly across different regions and demographics (i.e. skewing the data towards more developed geographic regions, or skewing away from geographic regions that use a platform other than Spotify mainly).
Another ethical consideration in our study is the fair/respectful representation of different cultural groups. There are several potential challenges here:

- The risk of oversimplification/overgeneralization when categorizing music by ethnic or cultural groups
- Potential microaggressions in the process of cultural classification
- Uneven representation of different regions and ethnic groups in our dataset
- Accurately representing multicultural influences in music is difficult

While some limitations are inherent to our data source and cannot be fully resolved, we took measures to address them. In terms of the data itself, we followed Spotify’s terms of service strictly and were transparent about data limitations in our analysis. For cultural representation, we used respectful, clear criteria for cultural categorization while acknowledging that our categories are simplifications and assigned arbitrarily. These steps are not perfect, but we ensured our research is conducted and presented responsibly with clear acknowledgment of its limitations/scope.


# Discussion and Conclusion

Music has been a form of art that is an essential part of cultures with diverse purposes, in which music reflects different cultures with their unique music styles, characteristics (e.g. instrumentation, rhythm, melody, and vocals.) Music genre categorization system has assisted people to understand the diversity and traits of each genre, while reflecting intercultural interactions, increased by globalization, and history of the evolution of genres.
The purpose of the research was to find out patterns of systemic changes in audio characteristics due to the adoption of each genre of music by separate ethnic groups as well as the presence of consistent patterns in modification for each culture.
Our hypothesis was that music genres would display statistically significant differences in measurable audio characteristics when they are adopted by different ethnic groups – groups of people identify themselves with common factors (e.g. language, cultural, practices, and history) – while each ethnic group showing consistent patterns on their ways of modifying features.

The dataset we used is “EveryNoise database with spotify features” (last_fm.db), which contained 764,429 songs collected through the Spotify API; the data included key metrics–such as danceability, instrumentalness, acousticness, energy, tempo, and valence– and rich metadata about artists and genres, which contains cultural indicators (such as artists’ country of origin and genre names). Before analysis, we cleaned the genre labels to extract cultural/regional indicators (e.g. “korean,” “swedish,” and “dutch”), and standardized artist location data. Due to the large size (approx.. 600MB) of the database causing problems for execution times, we also generated a subset of the data – a 10% random sample– to perform our analysis, then convert it into a .csv file for easier handling.

When looking at the genre’s musical factors and their variations between countries, we can see that there is enough variation between them to be considered significant. With the exceptions of danceability and tempo, we can see that countries usually have variations in musical factors between them. Even outside of genres we see that countries have differing musical factor values that imply an influence in specific genres because of the geographical location.

All three of our genres in hip-hop, rock, and pop illustrate the variations in musical factors between different countries. We can see that Chinese pop music has a much lower average valence than the other countries which correlates to the more melancholic style of the music. Similarly, Japanese rock has a higher valence and energy than its counterparts which can be heard in the thrilling Anime OST's and more upbeat style of rock music popular in Japan

Its through both results and personal experience bolstering analysis that we can conclude that geographical location, and therefore culture, does have a significant impact on the musical style of specific genres.

## Limitations

Sampling bias may have occurred because of the demographics and accessibility of Spotify. The choropleth map of music genre distribution across the countries from our EDA showed the potential sampling bias in our dataset since actual musical diversity was not fully reflected on the dataset, while the dataset may reflect Spotify’s market availability. Due to the nature of the data that Spotify users are mostly distributed in Europe, North America, and Latin America, the sub-genre categorization and available music on Spotify are likely to poorly reflect the music and genres of the rest of the world. Moreover, various degrees of the accessibility of technology and popularity of the music stream services due to geological locations and socioeconomic status may have affected the music available on Spotify.


## Future Direction

For future research endeavours, we could expand analysis to include a broader range of countries and music genres, and even focus on non-Western music to capture a more representative view of global musical diversity. Deeper exploration of niche sub-genres, especially within the electronic music genre, could provide more insight into regional and cultural musical preferences. Furthermore, socioeconomic factors like age, gender, and economic class could add depth into our understanding of how cultural and societal influences can affect music consumption and genre popularity. Longitudinal studies can also be carried out over time to consider the evolution of musical trends, and emerging shifts in global musical preference along with technological influence on music production can be better analyzed.

## Societal Impact

Our data and conclusion indicates that culture and geography have an impact on music and can illustrate how important things like cultural influence are on genres. It also solidifies ideas that ethnic subgenres of music are independent of each other and cannot easily be lumped into a broad definition.

## Team Contributions

*   Pia: Abstract, background and prior work, correlation matrix of audio features analysis, hip-hop and pop music audio features distribution by country analysis, video editing (putting audio and slides together), video upload

* Jiwoo: background and prior work, ethics/limitation content, Discussion & conclusion, revision of paper, video presentation slides, Ideas, leadership, writing, organizing meetings

* Asher: hypothesis, EDA, general musical factor analysis, slides, video narration

* Kelsie: background and prior work, top 20 most common genres analysis, rock music audio features distribution by country analysis

* Jason: Initial project idea, research question, dataset gathering, dataset overview, data cleaning/wrangling & feature engineering, EDA charts/graphs/figures, genre choropleth map + analysis, "audio features by country" subplot figure, "feature distribution by country" figures, Kruskal-Wallis analysis